### ISSUE TITLE
Remove side-effects from convolution.convolve_fft

### ISSUE THREAD

**Issue by jamienoss** [astropy/astropy#8107]
[[src](https://github.com/astropy/astropy/blob/master/astropy/convolution/convolve.py#L777-L781)]

```python
# restore NaNs in original image (they were modified inplace earlier)
# We don't have to worry about masked arrays - if input was masked, it was
# copied
array[nanmaskarray] = np.nan
kernel[nanmaskkernel] = np.nan
```

**Comment by jamienoss** [comment]
In fact, this is not even correct! ``nanmaskarray`` is computed as ``~np.isfinite(array)`` ([src](https://github.com/astropy/astropy/blob/master/astropy/convolution/convolve.py#L615-L617)), and therefore any ``inf``` in the input array or kernel will be incorrectly replaced by ``Nan``.

This is a severe side effect, beyond the normal of the func returning before said values could be replaced.

**Comment by pulkit6559** [comment]
I'm new to astropy and merely a beginner, is it okay if i work on this?

**Comment by pllim** [comment]
@pulkit6559 , please checkout the issues labeled as "good first issue" first. Thank you for your interest!

**Comment by jamienoss** [comment]
@pulkit6559 I wasn't necessarily intending to work on this. The soln. mainly involves correctly copying the input arrays only when necessary. See https://github.com/astropy/astropy/pull/7503

One caveat, I wouldn't start until #https://github.com/astropy/astropy/issues/8114 is resolved, which I'm currently working on. ``convolve_fft()`` will be refactored in which case starting now might mean doing twice the work for you. Unless you finish first in which case it's twice for me, but I'm cool wit that 😉 

Also, I should note that I am not a maintainer of this package so have no actual official say when it comes to workload distribution. That being said, it is open src and, canonically, everyone is "free" to work on whatever they want. The only caveat is that someone else might beat you to the punch and/or do a better job. Either way, you are free to try.

**Comment by pulkit6559** [comment]
@pllim @jamienoss thankyou for the response, i'll try to tackle the easier ones first and then see if i could come up with something for this. :)

**Comment by jamienoss** [comment]
Sorry @pulkit6559 I had to go ahead and fix this for #8114.

**Comment by pulkit6559** [comment]
@jamienoss That's alright :), i was already occupied with #15492 :)
